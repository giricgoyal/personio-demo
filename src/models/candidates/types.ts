export type CANDIDATE = {
    application_date?: string
    birth_date?: string
    email?: string
    id?: number
    name?: string
    position_applied?: string
    status?: string
    year_of_experience?: number
}

export type ACTION = {
    type: string
    payload?: {
        isLoading?: boolean
        data?: Array<CANDIDATE>
    }
}

export type CANDIDATE_STATE = {
    candidates: {
        data?: Array<CANDIDATE>
        isLoading?: boolean
    }
}
