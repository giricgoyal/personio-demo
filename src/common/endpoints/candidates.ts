import { CANDIDATE } from 'src/models/candidates/types'
import { getData } from '../api'

export type CANDIDATE_RESPONSE = {
    data?: CANDIDATE[]
    error?: {
        code?: number
        message?: string
    }
}

export const getCandidates = (): Promise<CANDIDATE_RESPONSE> => getData('candidates')
