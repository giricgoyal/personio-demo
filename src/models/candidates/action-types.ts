export const FETCH_CANDIDATES = 'models/candidates/fetch'
export const FETCH_CANDIDATES_SET_LOADING = 'models/candidates/set-loading'
export const SET_CANDIDATES_DATA = 'models/candidates/set-data'
