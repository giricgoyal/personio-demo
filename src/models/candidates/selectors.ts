import { CANDIDATE, CANDIDATE_STATE } from './types'

const getState = (state: CANDIDATE_STATE) => state?.candidates

export const getCandidates = (state: CANDIDATE_STATE): Array<CANDIDATE> => getState(state)?.data ?? []

export const getIsLoading = (state: CANDIDATE_STATE): undefined | boolean => getState(state)?.isLoading

export const getError = (state: CANDIDATE_STATE): string => getState(state)?.error ?? ''
