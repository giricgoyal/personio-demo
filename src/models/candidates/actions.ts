import { FETCH_CANDIDATES, FETCH_CANDIDATES_SET_LOADING, SET_CANDIDATES_DATA, SET_ERROR } from './action-types'
import { CANDIDATE } from './types'

export const fetchCandidates = (): {
    type: string
} => ({
    type: FETCH_CANDIDATES,
})

export const setCandidatesData = (
    data: Array<CANDIDATE> | undefined,
): {
    type: string
    payload: {
        data: Array<CANDIDATE> | undefined
    }
} => ({
    type: SET_CANDIDATES_DATA,
    payload: {
        data,
    },
})

export const setIsLoading = (
    isLoading: boolean,
): {
    type: string
    payload: {
        isLoading: boolean
    }
} => ({
    type: FETCH_CANDIDATES_SET_LOADING,
    payload: {
        isLoading,
    },
})

export const setError = (
    error: string,
): {
    type: string
    payload: {
        error: string
    }
} => ({
    type: SET_ERROR,
    payload: {
        error,
    },
})
