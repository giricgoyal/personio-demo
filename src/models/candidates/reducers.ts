import { FETCH_CANDIDATES_SET_LOADING, SET_CANDIDATES_DATA } from './action-types'
import { ACTION, CANDIDATE } from './types'

export const initialState = {
    data: [],
    isLoading: false,
}

const candidates = (
    state = initialState,
    action: ACTION,
): {
    data: Array<CANDIDATE>
    isLoading: boolean
} => {
    const { type, payload } = action
    switch (type) {
        case SET_CANDIDATES_DATA: {
            return {
                ...state,
                ...payload,
            }
        }
        case FETCH_CANDIDATES_SET_LOADING: {
            return {
                ...state,
                ...payload,
            }
        }
        default: {
            return state
        }
    }
}

export default {
    candidates,
}
