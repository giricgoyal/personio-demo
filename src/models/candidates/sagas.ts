import { call, CallEffect, fork, put, PutEffect, takeLatest } from 'redux-saga/effects'
import { FETCH_CANDIDATES } from './action-types'
import { setCandidatesData, setError, setIsLoading } from './actions'
import { ACTION } from './types'
import { CANDIDATE_RESPONSE, getCandidates } from 'src/common/endpoints/candidates'

export function* fetchCandidates(): Generator<
    CallEffect<CANDIDATE_RESPONSE> | PutEffect<ACTION>,
    void,
    CANDIDATE_RESPONSE
> {
    try {
        yield put(setIsLoading(true))
        const result: CANDIDATE_RESPONSE = yield call(getCandidates)
        if (result.data) {
            yield put(setCandidatesData(result?.data))
            yield put(setError(''))
        } else if (result.error) {
            throw result.error
        }
    } catch (error) {
        yield put(setError('An error occured'))
    } finally {
        yield put(setIsLoading(false))
    }
}

function* fetchCandidatesSaga() {
    yield takeLatest(FETCH_CANDIDATES, fetchCandidates)
}

export default [fork(fetchCandidatesSaga)]
