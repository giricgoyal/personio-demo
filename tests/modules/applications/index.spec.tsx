import React from 'react'
import { makeTestStore, render } from 'tests/test-utils'
import Applications from 'src/modules/applications'
import { fetchCandidates } from 'src/models/candidates/actions'
import { fireEvent } from '@testing-library/react'

describe('modules/applications', () => {
    let container, getByTestId
    let store

    describe('When component is rendered first', () => {
        beforeEach(() => {
            store = makeTestStore()
            ;({ container } = render(<Applications />, { store }))
        })

        test('should render and match snapshot', () => {
            expect(container).toMatchSnapshot()
        })

        test('should call the fetchCandidates action', () => {
            expect(store.dispatch).toHaveBeenCalledWith(fetchCandidates())
        })
    })

    describe('When data is loading', () => {
        beforeEach(() => {
            store = makeTestStore({
                candidates: {
                    isLoading: true,
                },
            })
            ;({ container } = render(<Applications />, { store }))
        })

        test('should render and match snapshot', () => {
            expect(container).toMatchSnapshot()
        })
    })

    describe('When data is present in store', () => {
        beforeEach(() => {
            store = makeTestStore({
                candidates: {
                    data: [
                        {
                            name: 'John Doe',
                            years_of_experience: 23,
                            position_applied: 'writer',
                            application_date: '2019-10-10',
                        },
                        {
                            name: 'Jane Smith',
                            years_of_experience: 23,
                            position_applied: 'QA Engineer',
                            application_date: '2022-10-10',
                        },
                    ],
                },
            })
            ;({ container, getByTestId } = render(<Applications />, { store }))
        })

        test('should render and match snapshot', () => {
            expect(container).toMatchSnapshot()
        })

        describe('When sort is changed', () => {
            beforeEach(() => {
                fireEvent.change(getByTestId('select'), {
                    target: {
                        value: 'position_applied',
                    },
                })
            })

            test('should render and match snapshot', () => {
                expect(container).toMatchSnapshot()
            })
        })
    })
})
