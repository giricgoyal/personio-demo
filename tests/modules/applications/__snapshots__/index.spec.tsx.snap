// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modules/applications should render and match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="grid"
      >
        <div
          class="grid__header"
        >
          <div
            class="grid__header__left"
          >
            <div
              class="grid__header__left__title"
            >
              Applications
            </div>
            <div
              class="grid__header__left__data-count"
            >
              (
              0
              /
              0
              )
            </div>
          </div>
          <div
            class="grid__header__right"
          >
            <div>
              Sort By:
            </div>
          </div>
        </div>
        <div
          class="grid__table grid__table--grid-8"
        >
          <div
            class="grid__table__cell grid__table__cell--header"
          >
            Id
          </div>
          <div
            class="grid__table__cell grid__table__cell--header"
          >
            Name
          </div>
          <div
            class="grid__table__cell grid__table__cell--header"
          >
            Email
          </div>
          <div
            class="grid__table__cell grid__table__cell--header"
          >
            Birth Date
          </div>
          <div
            class="grid__table__cell grid__table__cell--header"
          >
            Application Date
          </div>
          <div
            class="grid__table__cell grid__table__cell--header"
          >
            Position Applied
          </div>
          <div
            class="grid__table__cell grid__table__cell--header"
          >
            Status
          </div>
          <div
            class="grid__table__cell grid__table__cell--header"
          >
            Years of experience
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="grid"
    >
      <div
        class="grid__header"
      >
        <div
          class="grid__header__left"
        >
          <div
            class="grid__header__left__title"
          >
            Applications
          </div>
          <div
            class="grid__header__left__data-count"
          >
            (
            0
            /
            0
            )
          </div>
        </div>
        <div
          class="grid__header__right"
        >
          <div>
            Sort By:
          </div>
        </div>
      </div>
      <div
        class="grid__table grid__table--grid-8"
      >
        <div
          class="grid__table__cell grid__table__cell--header"
        >
          Id
        </div>
        <div
          class="grid__table__cell grid__table__cell--header"
        >
          Name
        </div>
        <div
          class="grid__table__cell grid__table__cell--header"
        >
          Email
        </div>
        <div
          class="grid__table__cell grid__table__cell--header"
        >
          Birth Date
        </div>
        <div
          class="grid__table__cell grid__table__cell--header"
        >
          Application Date
        </div>
        <div
          class="grid__table__cell grid__table__cell--header"
        >
          Position Applied
        </div>
        <div
          class="grid__table__cell grid__table__cell--header"
        >
          Status
        </div>
        <div
          class="grid__table__cell grid__table__cell--header"
        >
          Years of experience
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
